File: code/greedy/T4026.py
class Solution:
    def maximumGap(self, skill: str, station: str) -> int:
        n = len(skill)
        suf = [0] * n
        # 从后往前找子序列
        j = len(station) - 1
        for i in range(n-1, 0, -1):
            while skill[i] != station[j]:
                j -= 1
            suf[i] = j
            j -= 1
        # 从前往后找前部分子序列
        pre = 0
        ans = 0
        for i in range(n-1):
            while skill[i] != station[pre]:
                pre += 1
            ans = max(ans, suf[i+1] - pre)
            pre += 1
        return ans
        

File: src/leetcode/greedy/T4026.java
import Math;

class Solution {
    public int maximumGap(String skill, String station) {
        int n = skill.length();
        int[] suf = new int[n];
        int j = station.length() - 1;
        for (int i = n - 1; i > 0; i--) {
            while (skill.charAt(i) != station.charAt(j)) {
                j--;
            }
            suf[i] = j;
            j--;
        }
        int ans = 0;
        int pre = 0;
        for (int i = 0; i < n - 1; i++) {
            while (skill.charAt(i) != station.charAt(pre)) {
                pre += 1;
            }
            ans = Math.max(ans, suf[i+1] - pre);
            pre += 1;
        }
        return ans;
    }
}